engine/src/main/java/org/hibernate/search/v6poc/engine/spi/BeanResolver.java

```diff
@@ -7,15 +7,47 @@
 package org.hibernate.search.v6poc.engine.spi;
 
 
+import org.hibernate.search.v6poc.util.SearchException;
+
 /**
  * @author Yoann Rodiere
  */
-public interface BeanResolver {
+public interface BeanResolver extends AutoCloseable {
+
+	/**
+	 * Release any internal resource created while resolving beans.
+	 * <p>
+	 * Provided beans will not be usable after a call to this method.
+	 *
+	 * @see AutoCloseable#close()
+	 */
+	@Override
+	void close();
 
 	<T> T resolve(Class<?> implementationType, Class<T> expectedClass);
 
 	<T> T resolve(String implementationName, Class<T> expectedClass);
 
-	<T> T resolve(BeanReference reference, Class<T> expectedClass);
+	default <T> T resolve(BeanReference reference, Class<T> expectedClass) {
+		String implementationName = reference.getName();
+		Class<?> implementationType = reference.getType();
+		boolean nameProvided = implementationName != null && !implementationName.isEmpty();
+		boolean typeProvided = implementationType != null;
+
+		if ( nameProvided && typeProvided ) {
+			throw new SearchException( "This bean resolver does not support"
+					+ " bean references using both a name and a type."
+					+ " Got both '" + implementationName + "' and '" + implementationType + "' in the same reference" );
+		}
+		else if ( nameProvided ) {
+			return resolve( implementationName, expectedClass );
+		}
+		else if ( typeProvided ) {
+			return resolve( implementationType, expectedClass );
+		}
+		else {
+			throw new SearchException( "Got an empty bean reference (no name, no type)" );
+		}
```

Comment:
I haven't reviewed the 5 PR back at the time but I'm a bit worried about us having something so complex as a default method of an SPI.

Reply:
In Search 5 We didn't, because we didn't have support for fetching beans by their name.

It's true that it's a bit complex for a default method. I can see two solutions:

1. I move this to an abstract base class, and let implementors extend from that, or not, as they see fit.
2. I add an internal layer on top of `BeanResolver`, let's say `BeanProvider`: other components would never use a `BeanResolver`, always a `BeanProvider`, and `BeanProvider` would be implemented in the engine.
`BeanResolver` would no longer expose a method accepting a `BeanReference`: just a method accepting a string reference (+ expected return type), a method accepting a type reference (+ expected return type), and a method accepting both (+ expected return type). `BeanResolver` would always expect all references to be non-null and non-empty. The third method would be optional, `BeanResolver` implementors being free to throw an exception, which would prevent users from using e.g. `@ValueBridgeBeanReference(name = "foo", type = MyType.class)`, with both attributes set. Not that currently it's always prevented, because of a limitation in how CDI integration is implemented in Hibernate ORM (CDI could handle that just fine...).
`BeanProvider` would just implement the methods currently in `BeanResolver` by delegating to the right method of a `BeanResolver`, which would include the code you referenced in your comment.

I think solution 2 would be safer in the long run, in particular because it decouples the bean resolver SPI from the other modules that use bean resolution.
Any preference?

Reply:
2. looks goods.

What also bugged me is that an ill-designed implementation in ORM leads the default implementation of our SPI to be quite limited.

Our SPI should be neutral IMHO. And as far as I understand your proposal for 2, it would be.